Introduce markdown linting

khiga8 opened the pull request:
Fixes: https://github.com/github/accessibility/issues/2550

This PR:
- Introduces markdown linting using GitHub's recommended markdownlint configs ([markdownlint-github](https://github.com/github/markdownlint-github)). These will help us enforce best accessibility practices in our docs.
- Configures the markdownlint VSCode markdown extension in `devcontainer` so that issues are flagged in the editor.
- Adds a CI job for markdown linting. I decided to separate the linting job to `eslint` and `markdownlint`.

khiga8 commented on .github/workflows/ci.yml:
```diff
@@ -26,7 +26,34 @@ jobs:
       - name: Lint
         run: npm run lint
-
+  markdown:
```
I copied this setup from PVC...

khiga8 commented:
Here are the issues that need to be resolved:

```
contributor-docs/adrs/adr-004-children-as-api.md:41 GH001/no-default-alt-text Images should set meaningful alternative text (alt text), and not use the macOS default screenshot filename or `Image`. [For image: <img width="373" alt="image" src="https://user-images.githubusercontent.com/1863771/144945223-70c4c800-5827-4985-9f18-0ab416eba058.png">] https://github.com/github/markdownlint-github/blob/main/docs/rules/GH001-no-default-alt-text.md
contributor-docs/adrs/adr-005-box-sx.md:28:1 MD030/list-marker-space Spaces after list markers [Expected: 1; Actual: 2] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md030
contributor-docs/adrs/adr-008-experimental-components.md:11 MD001/heading-increment/header-increment Heading levels should only increment by one level at a time [Expected: h3; Actual: h4] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md001
contributor-docs/adrs/adr-008-experimental-components.md:56 MD001/heading-increment/header-increment Heading levels should only increment by one level at a time [Expected: h3; Actual: h4] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md001
contributor-docs/adrs/adr-XXX-file-structure.md:15 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md040
contributor-docs/behaviors.md:20 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md040
contributor-docs/behaviors.md:108 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md040
contributor-docs/component-contents-api-patterns.md:5 MD001/heading-increment/header-increment Heading levels should only increment by one level at a time [Expected: h2; Actual: h3] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md001
contributor-docs/CONTRIBUTING.md:[13](https://github.com/primer/react/actions/runs/3802472579/jobs/6467983949#step:5:14)7 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md040
contributor-docs/CONTRIBUTING.md:215:1 MD053/link-image-reference-definitions Link and image reference definitions should be needed [Unused link or image reference definition: "classnames"] [Context: "[classnames]: https://www.npmj..."] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md053
contributor-docs/CONTRIBUTING.md:216:1 MD053/link-image-reference-definitions Link and image reference definitions should be needed [Unused link or image reference definition: "spread syntax"] [Context: "[spread syntax]: https://devel..."] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md053
contributor-docs/CONTRIBUTING.md:217:1 MD053/link-image-reference-definitions Link and image reference definitions should be needed [Unused link or image reference definition: "styled-system"] [Context: "[styled-system]: https://style..."] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md053
contributor-docs/CONTRIBUTING.md:218:1 MD053/link-image-reference-definitions Link and image reference definitions should be needed [Unused link or image reference definition: "table"] [Context: "[table]: https://jxnblk.com/st..."] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md053
contributor-docs/CONTRIBUTING.md:220:1 MD053/link-image-reference-definitions Link and image reference definitions should be needed [Unused link or image reference definition: "now"] [Context: "[now]: https://zeit.co/now"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md053
contributor-docs/principles.md:3 MD001/heading-increment/header-increment Heading levels should only increment by one level at a time [Expected: h2; Actual: h3] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md001
contributor-docs/testing.md:75 GH002/no-generic-link-text Avoid using generic link text like `Learn more` or `Click here` [For link: here] https://github.com/github/markdownlint-github/blob/main/docs/rules/GH002-no-generic-link-text.md
docs/content/anchoredPosition.mdx:[14](https://github.com/primer/react/actions/runs/3802472579/jobs/6467983949#step:5:15)8 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "Return value"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/core-concepts.md:11 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md040
docs/content/core-concepts.md:27 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md040
docs/content/deprecated/ActionList.mdx:13 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "Before"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/ActionList.mdx:27 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "After"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/ActionMenu.mdx:13 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "Before"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/ActionMenu.mdx:30 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "After"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/BorderBox.md:12 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "Before"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/BorderBox.md:18 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "After"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/Dropdown.md:51 MD001/heading-increment/header-increment Heading levels should only increment by one level at a time [Expected: h3; Actual: h4] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md001
docs/content/deprecated/DropdownMenu.mdx:12 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "Before"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/DropdownMenu.mdx:42 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "After"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/Flex.md:12 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "Before"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/Flex.md:22 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "After"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/Grid.md:12 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "Before"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/Grid.md:25 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "After"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/Position.md:12 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "Before"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/Position.md:24 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "After"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/SideNav.md:13 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "Before"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/SideNav.md:25 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "After"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/deprecated/SideNav.md:176 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md040
docs/content/Details.md:7 GH002/no-generic-link-text Avoid using generic link text like `Learn more` or `Click here` [For link: here] https://github.com/github/markdownlint-github/blob/main/docs/rules/GH002-no-generic-link-text.md
docs/content/FilteredSearch.md:31 MD001/heading-increment/header-increment Heading levels should only increment by one level at a time [Expected: h3; Actual: h4] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md001
docs/content/FilterList.md:25 MD001/heading-increment/header-increment Heading levels should only increment by one level at a time [Expected: h3; Actual: h4] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md001
docs/content/focusZone.mdx:130 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "Return value"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/NavList.mdx:325:84 MD037/no-space-in-emphasis Spaces inside emphasis markers [Context: "* Should `title... be required? *"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md037
docs/content/Overlay.mdx:71 GH002/no-generic-link-text Avoid using generic link text like `Learn more` or `Click here` [For link: here] https://github.com/github/markdownlint-github/blob/main/docs/rules/GH002-no-generic-link-text.md
docs/content/PageLayout.mdx:313 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "Links & Resources"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md036
docs/content/primer-theme.md:13 GH002/no-generic-link-text Avoid using generic link text like `Learn more` or `Click here` [For link: here] https://github.com/github/markdownlint-github/blob/main/docs/rules/GH002-no-generic-link-text.md
docs/content/ssr.mdx:9 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md040
docs/content/ssr.mdx:23 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md040
docs/content/Textarea.mdx:102 MD031/blanks-around-fences Fenced code blocks should be surrounded by blank lines [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md031
docs/content/Textarea.mdx:102 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md040
migrating.md:12:12 MD051/link-fragments Link fragments should be valid [Context: "[codemod](#codemods) available to upgrade component identifiers from [v3.x.x-beta](#3-0-0-beta)"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md051
migrating.md:14 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md040
migrating.md:33:12 MD051/link-fragments Link fragments should be valid [Context: "[codemod](#codemods) available to upgrade from [2.x.x-beta](#2-0-0-beta)"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md051
migrating.md:35 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md040
migrating.md:49:12 MD051/link-fragments Link fragments should be valid [Context: "[codemod](#codemods) available to upgrade from [1.x.x-beta](#1-0-0-beta)"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md051
migrating.md:51 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md040
migrating.md:61 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md040
migrating.md:248:1 MD053/link-image-reference-definitions Link and image reference definitions should be needed [Unused link or image reference definition: "themeget"] [Context: "[themeget]: http://jxnblk.com/..."] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md053
README.md:43 GH002/no-generic-link-text Avoid using generic link text like `Learn more` or `Click here` [For link: here] https://github.com/github/markdownlint-github/blob/main/docs/rules/GH002-no-generic-link-text.md
README.md:55:1 MD053/link-image-reference-definitions Link and image reference definitions should be needed [Unused link or image reference definition: "styled-components"] [Context: "[styled-components]: https://w..."] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md053
README.md:56:1 MD053/link-image-reference-definitions Link and image reference definitions should be needed [Unused link or image reference definition: "primer css"] [Context: "[primer css]: https://github.c..."] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md053
README.md:57:1 MD053/link-image-reference-definitions Link and image reference definitions should be needed [Unused link or image reference definition: "flash of unstyled content"] [Context: "[flash of unstyled content]: h..."] https://github.com/DavidAnson/markdownlint/blob/v0.26.2/doc/Rules.md#md053
```

khiga8 commented:
`no-duplicate-header` is an issue that appears a lot!!

I think we could configure this rule to be `sibling_only: true` if https://github.com/primer/doctocat/issues/527 gets resolved. (docs at [MD024](https://github.com/DavidAnson/markdownlint/blob/main/doc/md024.md))

khiga8 commented:
Can I get a first-pass review from @primer/react-reviewers on these configs?

broccolinisoup commented on package.json:
```diff
@@ -45,6 +45,7 @@
     "install:docs": "(cd docs && npm install --legacy-peer-deps)",
     "lint": "eslint '**/*.{js,ts,tsx,md,mdx}' --max-warnings=0",
     "lint:fix": "npm run lint -- --fix",
+    "markdownlint": "markdownlint-cli2 \"**/*.{md,mdx}\" \"!.github\" \"!.changeset\" \"!node_modules\" \"!CHANGELOG.md\"",
```
We have `node_modules` in a few more places such as `docs/` so do you think would it makes sense to extend the coverage like this?
 
```suggestion
    "markdownlint": "markdownlint-cli2 \"**/*.{md,mdx}\" \"!.github\" \"!.changeset\" \"!**/node_modules\" \"!CHANGELOG.md\"",
```

joshblack replied:
Small name suggestion, let me know if it makes sense! 

```suggestion
    "lint:md": "markdownlint-cli2 \"**/*.{md,mdx}\" \"!.github\" \"!.changeset\" \"!node_modules\" \"!CHANGELOG.md\"",
```

broccolinisoup commented on package.json:
```diff
@@ -45,6 +45,7 @@
     "install:docs": "(cd docs && npm install --legacy-peer-deps)",
     "lint": "eslint '**/*.{js,ts,tsx,md,mdx}' --max-warnings=0",
```
We have a set of rules for the `md` and `mdx` files already [on the eslint config.](https://github.com/primer/react/blob/main/.eslintrc.json#L125) I am not sure if there would be any conflict between them? Or what do you suggest for us to do here?

khiga8 replied:
oooh interesting! I don't think `eslint` runs the same rulesets as markdownlint. do you know what rules are running on markdown?

broccolinisoup reviewed (commented):
👋🏼  @khiga8 Thank you so much for this PR 🙏🏼 I left a couple of comments, let us know what you think!

khiga8 commented on .markdownlint-cli2.cjs:
```diff
@@ -0,0 +1,30 @@
+// Rules we want to turn on but currently have too many violations
+const rulesToEnforce = {
+  'fenced-code-language': false,
+  'no-duplicate-header': false, // Fix https://github.com/primer/doctocat/issues/527, then set this rule to `siblings_only: true`
```
@primer/react-reviewers FYI,`no-duplicate-header` should be turned on because doctocat messes up the anchor links when headings have the same text. See linked issue. I keep it off for now because there's too many address.

Ideally, the issue is fixed in doctocat so we don't have to set this rule to be strict.

khiga8 commented on contributor-docs/adrs/adr-004-children-as-api.md:
```diff
@@ -38,7 +39,7 @@ On the other hand, if we want consumers to have more control over children, a co
 With React, `children` is the out-of-the-box way for putting [phrasing content](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#phrasing_content) inside your component. By using `children` instead of our own custom prop, we can make the API “predictable” for its consumers.
 
-<img width="373" alt="image" src="https://user-images.githubusercontent.com/1863771/144945223-70c4c800-5827-4985-9f18-0ab416eba058.png">
+<img width="373" alt="image" src="https://user-images.githubusercontent.com/1863771/144945223-70c4c800-5827-4985-9f18-0ab416eba058.png"> <!-- markdownlint-disable-line no-default-alt-text -->
```
@primer/react-reviewers This is a TODO!

khiga8 commented on contributor-docs/component-contents-api-patterns.md:
```diff
@@ -2,7 +2,7 @@
 Consider a React component that renders a list of items. Here are two possible APIs that component might expose, both achieving an equivalent result.
 
-### A: Contents passed as React children
+### A: Contents passed as React children <!-- markdownlint-disable-line heading-increment -->
```
@primer/react-reviewers This is a TODO! I'm not sure what the heading structure should be. Feel free to suggest something in this PR to fix this!

khiga8 commented on contributor-docs/adrs/adr-005-box-sx.md:
````diff
@@ -10,44 +10,44 @@ In Primer React and consuming applications, we use many different patterns for c
-    ```
-
-    [Show full code example →](https://github.com/primer/react/pull/2019/files?diff=split&w=0)
+   ```tsx
````
Errr I think prettier ran for these files and updated it, but not sure. @primer/react-reviewers is this expected?

khiga8 commented on .markdownlint-cli2.cjs:
```diff
@@ -0,0 +1,30 @@
+  'no-duplicate-header': false, // Fix https://github.com/primer/doctocat/issues/527, then set this rule to `siblings_only: true`
+}
+// Rules we don't care to enforce (usually stylistic)
+const rulesToNotEnforce = {
+  'line-length': false,
+  'blanks-around-headings': false,
+  'blanks-around-lists': false,
+  'no-trailing-spaces': false,
+  'no-multiple-blanks': false,
+  'no-trailing-punctuation': false,
+  'single-trailing-newline': false,
+  'ul-indent': false,
+  'no-hard-tabs': false,
+  'first-line-heading': false,
+  'no-space-in-emphasis': false,
+  'blanks-around-fences': false,
+}
```
These are stylistic rules that I don't think Primer team cares to enforce for now, but not sure. Feel free to update this.

khiga8 commented:
@primer/react-reviewers This is ready for another review.

For the existing violations listed in https://github.com/primer/react/pull/2726#issuecomment-1367517548, I addressed some of them. You can go through them by commit.
Two of the rules, there were too many of to address immediately so I turned them off. 

In particular, I think `no-duplicate-header` should be turned on based on how doctocat currently renders heading anchors, but https://github.com/primer/doctocat/issues/527 would alleviate the need.

joshblack commented on .github/workflows/ci.yml:
```diff
@@ -26,7 +26,22 @@ jobs:
+      - name: Install dependencies
+        run: npm ci
+
+      - name: Lint
+        run: npm run markdownlint
```
Could we add this to the `lint` job as a step after `npm run lint`? Would be great to keep it in the same job for linting, something like:

```
      - name: Lint JavaScript
        run: npm run lint
      - name: Lint markdown
        run: npm run markdownlint
```

khiga8 replied:
Sure, I can do that! I got [feedback](https://github.com/primer/design/pull/345#discussion_r1055753994) in the `primer/design` repo that it's more clear when there are separate linting jobs for error clarity, but we can keep it together if you prefer. I don't have a strong opinion.

khiga8 replied:
Updated!

joshblack reviewed (approved):
Thanks for doing this! Left a comment for the CI job, let me know what you think about it 👀 Would be great to include, if possible.

khiga8 commented:
note to self to update the `CONTRIBUTING.md`

khiga8 commented:
Hi @joshblack! I addressed your review suggestions in https://github.com/primer/react/pull/2726/commits/d1f87d2bf606e2dfb645b493aceb0cb1d8425608. 

I additionally updated the `CONTRIBUTING.md` file in https://github.com/primer/react/pull/2726/commits/572c744a9b5cfbdba3f854597d7731680d9f97e2.

joshblack commented:
Thanks for doing this @khiga8! 🙌
